### PR TITLE
Feature/deb 99115

### PR DESCRIPTION
**Pull request by wenpengxu**

### CODE REVIEW

**Comment by tmessina** [comment]
@SwatiJainTT  please take a look as well. did we want to/need to change the algo enums as well?

File: dotnet/tt-net-sdk/marketdata/InstrumentMarketDataCache.cs [hunk 1]
```diff
@@ -162,7 +162,7 @@ internal void HandleUpdate(List<MarketDataUpdate> updates)
                     BidCounterPartyId  .ApplyValue(update.TradeData.BidCounterPartyId);
                     AskCounterPartyId  .ApplyValue(update.TradeData.AskCounterPartyId);
 
-                    if( update.TradeData.TradingStatus != tt_net_sdk.TradingStatus.Unknown )
+                    if( update.TradeData.TradingStatus != tt_net_sdk.TradingStatus.NotSet )
```
Comment (tmessina) [inline dotnet/tt-net-sdk/marketdata/InstrumentMarketDataCache.cs]: Should this be !unknown AND !notset ? I do not think we want to set the status to unknown if we didn't before

File: dotnet/tt-net-sdk/messages_generated/AlgoParameterUpdate.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        public NullableUInt64 time_sent        {get;set;}
+        public UserParameters user_parameters  {get;set;}
+        public NullableString exch_order_assoc {get;set;}
+
+        public override string ToString()
+        {
+            return AsJson();
+        }
+
+    }
```
Comment (tmessina) [inline dotnet/tt-net-sdk/messages_generated/AlgoParameterUpdate.cs]: Why was this file added for this jira?

File: dotnet/tt-net-sdk/messages_generated/AlgoSideChannelRequest.cs [hunk 3]
```diff
@@ -28,13 +28,13 @@ namespace tt_net_sdk.messages.algo_server
...
-        public NullableUInt64             account_id {get;set;}
-        public NullableString             server_id  {get;set;}
-        public AlgoSideChannelRequestType type       {get;set;}
-        public string[]                   cmd_args   {get;set;}
+        public NullableUInt64              cl_ord_id  {get;set;}
+        public NullableString              order_id   {get;set;}
+        public NullableUInt64              user_id    {get;set;}
+        public NullableUInt64              account_id {get;set;}
+        public NullableString              server_id  {get;set;}
+        public AlgoSideChannelRequestType? type       {get;set;}
```
Comment (tmessina) [inline dotnet/tt-net-sdk/messages_generated/AlgoSideChannelRequest.cs]: Why do this messages now have a ? for the enums? Wouldn't the enums be NotSet if not there when extracted?

File: dotnet/tt-net-sdk/orders_fills/ExecutionBook.cs [hunk 4]
```diff
@@ -584,7 +585,7 @@ private void ProcessCancelReject(OrderCancelReject cancelReject)
                     orders_.TryGetValue(orderId, out holder);
                     // make sure this is actually the reponse to the message we sent by checking the application id
                     string erAppId = ServiceUtils.s(cancelReject.appl_id, string.Empty);
-                    MarketId mrk = ServiceUtils.toEnum<MarketId>(cancelReject.market_id, MarketId.INVALID);
+                    MarketId mrk = ServiceUtils.toEnum<MarketId>(cancelReject.market_id, MarketId.INVALID, MarketId.NotSet);
                     if (mrk == MarketId.INVALID && holder?.CurrentOrder != null)
```
Comment (tmessina) [inline dotnet/tt-net-sdk/orders_fills/ExecutionBook.cs]: will this still catch the algo case or should be we checking for not set instead of invalid? I think it should be NotSet 
 @SwatiJainTT

File: dotnet/tt-net-sdk/orders_fills/ExecutionBook.cs [hunk 5]
```diff
@@ -760,7 +761,7 @@ internal void ProcessExecutionReport(ExecutionReport executionReport, bool isFul
                     // make sure this is actually the reponse to the message we sent by checking the application id
                     orders_.TryGetValue(orderId, out holder);
                     string erAppId = ServiceUtils.s(executionReport.appl_id, string.Empty);
-                    MarketId mrk = ServiceUtils.toEnum<MarketId>(executionReport.market_id, MarketId.INVALID);
+                    MarketId mrk = ServiceUtils.toEnum<MarketId>(executionReport.market_id, MarketId.INVALID, MarketId.NotSet);
                     if (mrk == MarketId.INVALID && holder?.CurrentOrder != null)
```
Comment (tmessina) [inline dotnet/tt-net-sdk/orders_fills/ExecutionBook.cs]: Same comment as above -- think this should be NotSet. @SwatiJainTT  to confirm

File: dotnet/tt-net-sdk/orders_fills/OrderProfile.cs [hunk 6]
```diff
@@ -69,7 +69,7 @@ public OrderAction Action
                 else if (action_ == OrderAction.Resubmit || action_ == OrderAction.Resume)
                     orderRep_.execInst_ = ExecInst.Work;
                 else
-                    orderRep_.execInst_ = ExecInst.Unknown;
+                    orderRep_.execInst_ = ExecInst.Unknown;  // can't be NotSet
```
Comment (tmessina) [inline dotnet/tt-net-sdk/orders_fills/OrderProfile.cs]: Why can't it be NotSet -- please update the comment

File: dotnet/tt-net-sdk/products_and_instruments/InstrumentDetail.cs [hunk 7]
```diff
@@ -62,9 +62,9 @@ public Int64 LotSize
             }
         }
         public string Name                  => name_;
-        public OptionType OptionType        => ServiceUtils.toEnum<OptionType>(instrumentSource_.optioncodeid, OptionType.NotSet);
+        public OptionType OptionType        => ServiceUtils.toEnum<OptionType>(instrumentSource_.optioncodeid, OptionType.NotSet, OptionType.NotSet);
         public Int64 PriceDisplayDecimals   => ServiceUtils.l(instrumentSource_.pricedisplaydecimals);
-        public PriceDisplayType PriceDisplayType => ServiceUtils.toEnum<PriceDisplayType>(instrumentSource_.pricedisplaytypeid, PriceDisplayType.TtPriceDisplayNormal);
+        public PriceDisplayType PriceDisplayType => ServiceUtils.toEnum<PriceDisplayType>(instrumentSource_.pricedisplaytypeid, PriceDisplayType.TtPriceDisplayNormal, PriceDisplayType.NotSet);
```
Comment (tmessina) [inline dotnet/tt-net-sdk/products_and_instruments/InstrumentDetail.cs]: Wondering if maybe this should default to TtPriceDisplayNormal when it is not there since PDS might be assuming that default

File: dotnet/tt-net-sdk/products_and_instruments/InstrumentDetail.cs [hunk 8]
```diff
@@ -36,13 +36,13 @@ public class InstrumentDetails
...
         public string RollingAlias      => ServiceUtils.s(instrumentSource_.rollingalias, string.Empty);
-        public ComboCode ComboCode      => ServiceUtils.toEnum<ComboCode>(instrumentSource_.combotypeid, ComboCode.NotSet);
+        public ComboCode ComboCode      => ServiceUtils.toEnum<ComboCode>(instrumentSource_.combotypeid, ComboCode.NotSet, ComboCode.NotSet);
         public CurrencyCode Currency
         {
             get
             {
                 return ServiceUtils.toEnum<CurrencyCode>(instrumentSource_.productcurrencytype,
-                            TTAPI.Instance.Options.PrimaryCurrency);
+                            TTAPI.Instance.Options.PrimaryCurrency, CurrencyCode.NotSet);
```
Comment (tmessina) [inline dotnet/tt-net-sdk/products_and_instruments/InstrumentDetail.cs]: Interesting - we said before if there was not currency, we would default it to the primary currency set by the user in the TTAPI options. We have changed that to not be notset if not there. I think we might want the currency to be the primary one if not there. @mjmessina-tt  thoughts?

File: dotnet/tt-net-sdk/products_and_instruments/Product.cs [hunk 9]
```diff
@@ -52,14 +52,14 @@ private Product(MarketId mrkId, ProductType type, UInt64 id)
...
+            MarketId mKey = ServiceUtils.toEnum<MarketId>(prodRecord.marketid, tt_net_sdk.MarketId.INVALID, MarketId.NotSet);
             market_     = TTAPI.PDSService.GetMarket(mKey);
             productKey_ = new ProductKey(mKey,
-                ServiceUtils.toEnum<ProductType>(prodRecord.typeid, Market.IsSynthetic(mKey) ? ProductType.Synthetic : ProductType.NoSecurityType),
+                ServiceUtils.toEnum<ProductType>(prodRecord.typeid, Market.IsSynthetic(mKey) ? ProductType.Synthetic : ProductType.NoSecurityType, ProductType.NotSet),
                 prodRecord.symbol, id_);
             description_ = ServiceUtils.s(prodRecord.name, string.Empty);
             currency_ = new Currency(ServiceUtils.toEnum<CurrencyCode>(prodRecord.currencyid,
-                                        TTAPI.Instance.Options.PrimaryCurrency));
+                                        TTAPI.Instance.Options.PrimaryCurrency, CurrencyCode.NotSet));
```
Comment (tmessina) [inline dotnet/tt-net-sdk/products_and_instruments/Product.cs]: See currency comment above.

File: dotnet/tt-net-sdk/products_and_instruments/ProductDataService.cs [hunk 10]
```diff
@@ -516,7 +516,7 @@ private string GetShortDay(DateTime dt)
             {
                 foreach (var ms in msg.market_level_schedules)
                 {
-                    MarketId mrkId = ServiceUtils.toEnum<MarketId>(ms.market_id, MarketId.INVALID);
+                    MarketId mrkId = ServiceUtils.toEnum<MarketId>(ms.market_id, MarketId.INVALID, MarketId.NotSet);
                     if ((mrkId != MarketId.INVALID) && (ms.events != null) && (ms.events.Length > 0))
```
Comment (tmessina) [inline dotnet/tt-net-sdk/products_and_instruments/ProductDataService.cs]: this also probably be checking for not set

File: dotnet/tt-net-sdk/products_and_instruments/ProductDataService.cs [hunk 11]
```diff
@@ -1294,7 +1294,7 @@ public SortedSet<ProductType> GetProductTypesForMarket(Market mrk)
                     {
                         foreach (var pt in msg.producttypes)
                         {
-                            ProductType type = ServiceUtils.toEnum<ProductType>(pt.id, ProductType.Invalid);
+                            ProductType type = ServiceUtils.toEnum<ProductType>(pt.id, ProductType.Invalid, ProductType.NotSet);
                             if ((type != ProductType.Invalid) && !productTypes.Contains(type))
```
Comment (tmessina) [inline dotnet/tt-net-sdk/products_and_instruments/ProductDataService.cs]: Check for not set as well.

File: dotnet/tt-net-sdk/marketdata/InstrumentMarketDataCache.cs (hunk 1 above)
Comment (wenpengxu) [reply dotnet/tt-net-sdk/marketdata/InstrumentMarketDataCache.cs]: changed to
 if (update.TradeData.TradingStatus != tt_net_sdk.TradingStatus.NotSet && update.TradeData.TradingStatus != tt_net_sdk.TradingStatus.Unknown)
   ...

File: dotnet/tt-net-sdk/messages_generated/AlgoParameterUpdate.cs (hunk 2 above)
Comment (wenpengxu) [reply dotnet/tt-net-sdk/messages_generated/AlgoParameterUpdate.cs]: I can remove it.  I add it because it is a new generated file.

File: dotnet/tt-net-sdk/messages_generated/AlgoSideChannelRequest.cs (hunk 3 above)
Comment (wenpengxu) [reply dotnet/tt-net-sdk/messages_generated/AlgoSideChannelRequest.cs]: they are generated by our generator. I don't know if it caused by our generator or proto file

File: dotnet/tt-net-sdk/orders_fills/OrderProfile.cs (hunk 6 above)
Comment (wenpengxu) [reply dotnet/tt-net-sdk/orders_fills/OrderProfile.cs]: was thinking All OrderReport should have a valid Execution value.  after review the code again, I changed it to NotSet

Comment (tmessina) [reply dotnet/tt-net-sdk/orders_fills/OrderProfile.cs]: It could be notset before the order is submitted so it is probably better to be notset. thx

File: dotnet/tt-net-sdk/messages_generated/AlgoSideChannelRequest.cs (hunk 3 above)
Comment (tmessina) [reply dotnet/tt-net-sdk/messages_generated/AlgoSideChannelRequest.cs]: @SwatiJainTT  why were these values not ? before?

File: dotnet/tt-net-sdk/products_and_instruments/InstrumentDetail.cs (hunk 7 above)
Comment (wenpengxu) [reply dotnet/tt-net-sdk/products_and_instruments/InstrumentDetail.cs]: changed to TtPriceDisplayNormal .

File: dotnet/tt-net-sdk/products_and_instruments/ProductDataService.cs (hunk 10 above)
Comment (wenpengxu) [reply dotnet/tt-net-sdk/products_and_instruments/ProductDataService.cs]: yes. changed.

File: dotnet/tt-net-sdk/products_and_instruments/ProductDataService.cs (hunk 11 above)
Comment (wenpengxu) [reply dotnet/tt-net-sdk/products_and_instruments/ProductDataService.cs]: changed.

File: dotnet/tt-net-sdk/products_and_instruments/InstrumentDetail.cs (hunk 8 above)
Comment (wenpengxu) [reply dotnet/tt-net-sdk/products_and_instruments/InstrumentDetail.cs]: changed

File: dotnet/tt-net-sdk/products_and_instruments/Product.cs (hunk 9 above)
Comment (wenpengxu) [reply dotnet/tt-net-sdk/products_and_instruments/Product.cs]: done.
